File: frontend/src/types/events.ts
import type { CustomerResponse } from './customers';
import type { OrganizationResponse } from './organizations';

export interface StaffInviteDetails {
  specificStaffIds?: number[]; // Optional: list of specific staff to invite
  openInvitation: boolean; // True if invitation is open to any staff
  numberOfStaffNeeded: number; // Required: number of staff slots for the event
}

export interface EventInventoryItem {
  inventoryId: number;
  itemName: string;
  quantity: number;
}

export interface CreateEventRequest {
  title: string;
  date: string; // ISO 8601 date string (e.g., "YYYY-MM-DD")
  time: string; // ISO 8601 time string (e.g., "HH:MM:SS")
  durationMinutes: number;
  location?: string;
  description?: string;
  inventoryItems: EventInventoryItem[];
  customerId?: number; // Optional: Single customer ID
  staffInvites: StaffInviteDetails; // Details for staff invitation
  organizationId?: number;
}

export interface RsvpStatus {
  userId: number;
  displayName: string;
  email: string;
  availability: string; // "available" or "unavailable"
}

export interface EventResponse {
  eventId: number | string;
  title: string;
  date: string;
  time: string;
  durationMinutes: number;
  location?: string;
  description?: string;
  inventoryItems?: EventInventoryItem[];
  customer?: CustomerResponse;
  rsvps?: RsvpStatus[];
  organization?: OrganizationResponse;
  openInvitation?: boolean;
  numberOfStaffNeeded?: number;
  isPlaceholder?: boolean;
}

export interface UpdateEventRequest {
  title?: string;
  date?: string;
  time?: string;
  durationMinutes: number;
  location?: string;
  description?: string;
  inventoryItems?: EventInventoryItem[];
  customerId?: number;
  staffInvites?: StaffInviteDetails;
  organizationId?: number;
}

export interface RsvpRequest {
  availability: string; // "available" or "unavailable"
}

export interface VirtualEvent {
  id: string; // Unique ID for virtual event
  title: string;
  start: Date;
  end: Date;
  isVirtual: true;
}

export type CalendarEvent = (EventResponse | VirtualEvent) & { start: Date; end: Date; };

File: backend/data/src/main/kotlin/alphainterplanetary/teven/data/event/Events.kt
package alphainterplanetary.teven.data.event

import alphainterplanetary.teven.data.customer.Customers
import alphainterplanetary.teven.data.organization.Organizations
import org.jetbrains.exposed.v1.core.Table

object Events : Table() {
  val id = integer("id").autoIncrement()
  val title = varchar("title", 255)
  val date = varchar("date", 255)
  val time = varchar("time", 255)
  val location = varchar("location", 255).nullable()
  val description = text("description").nullable()
  val customerId = integer("customer_id").references(Customers.id).nullable()
  val organizationId = integer("organization_id").references(Organizations.id)
  val openInvitation = bool("open_invitation")
  val numberOfStaffNeeded = integer("number_of_staff_needed")
  val durationMinutes = integer("duration_minutes").default(120)

  override val primaryKey = PrimaryKey(id)
}